File: Source/Orca/DPrep/DPrep/bfile.cpp
//--------------------------------------------------------------------
// file bfile.cpp
//
// Stephen D. Bay
// Copyright 2003
//--------------------------------------------------------------------

#include <iostream>
#include <fstream>
//#include <sstream>
#include <strstream>
#include <string>
#include <vector>
#include <algorithm>

#include <stdio.h>
#include <math.h>

#include <stdlib.h>

#include "bfile.h"
#include "misc.h"
#include "database.h" // for tokenize





//--------------------------------------------------------------------
// BFile 
//--------------------------------------------------------------------
BFile::BFile(string file, float missing_r, int missing_d) : infile_(0)
{
  data_file_ = file;
	missing_r_ = missing_r;
	missing_d_ = missing_d;

	set_file_ptr(file);
	read_header();
};


//--------------------------------------------------------------------
// ~BFile
//--------------------------------------------------------------------
BFile::~BFile()
{
	if (infile_ != 0) {
		infile_->close();
		delete infile_;
	}
};


//--------------------------------------------------------------------
// scale_01 
//
// Scales real values to the range [0,1].
//--------------------------------------------------------------------
void BFile::scale_01(string file,vector<real> &max, vector<real> &min) 
{

	//-------------------------------
	// open file for writing
	//
	ofstream outfile (file.c_str(),ios::out|ios::binary);
	if ( ! outfile ) {
		cerr << "Error: cannot open file " << file << endl;
		exit(0);
	};
	
	// write header information
	outfile.write((const char *) &records_,sizeof(int));
	outfile.write((const char *) &rvars_,sizeof(int));
	outfile.write((const char *) &dvars_,sizeof(int));

	vector<real> range(rvars_,0);

	get_max_min(max,min);
	for (int i=0;i<max.size();i++) {
		range[i] = max[i] - min[i];
	};

	cout << "  Scaling data file" << endl;
	//--------------------------------
	// read in file and scale it
	//
	seek_position(0);

	vector<real> R(rvars_,0);
	vector<integer> D(dvars_,0);
	vector<real> Rscale(rvars_,0);
	int ID;
	while (index_ < records_) {
		get_next(ID,R,D);
		for (int i=0;i<R.size();i++) {
			if (R[i] == missing_r_) {
				Rscale[i] = missing_r_;
			}
			else if (range[i] != 0) {
				Rscale[i] = (R[i] - min[i]) / range[i]; 
			}
			else {
				Rscale[i] = 0;
			}
		}
		outfile.write((const char *) &ID,sizeof(int));
		outfile.write((const char *) &Rscale[0],sizeof(real)*rvars_);
		outfile.write((const char *) &D[0],sizeof(integer)*dvars_);
	};

	outfile.close();
};


//--------------------------------------------------------------------
// scale_std
//
// Scales real values to standard deviations from the mean.
//--------------------------------------------------------------------
void BFile::scale_std(string file, vector<real> &mean, vector<real> &std) 
{

	//-------------------------------
	// open file for writing
	//
	ofstream outfile (file.c_str(),ios::out|ios::binary);
	if ( ! outfile ) {
		cerr << "Error: cannot open file " << file << endl;
		exit(0);
	};
	
	// write header information
	outfile.write((const char *) &records_,sizeof(int));
	outfile.write((const char *) &rvars_,sizeof(int));
	outfile.write((const char *) &dvars_,sizeof(int));

	cout << "  Scaling data file" << endl;
	//--------------------------------
	// read in file and scale it
	//
	seek_position(0);

	vector<real> R(rvars_,0);
	vector<integer> D(dvars_,0);
	vector<real> Rscale(rvars_,0);
	int ID;
	while (index_ < records_) {
		get_next(ID,R,D);
		for (int i=0;i<R.size();i++) {
			if (R[i] == missing_r_) {
				Rscale[i] = missing_r_;
			}
			else if (std[i] != 0) {
				Rscale[i] = (R[i] - mean[i]) / std[i]; 
			}
			else {
				Rscale[i] = 0;
			}
		}
		outfile.write((const char *) &ID,sizeof(int));
		outfile.write((const char *) &Rscale[0],sizeof(real)*rvars_);
		outfile.write((const char *) &D[0],sizeof(integer)*dvars_);
	};

	outfile.close();
};





//--------------------------------------------------------------------
// get_max_min
//
// Makes a pass through the data and calculates the max and min values
// for real variables.
//--------------------------------------------------------------------
void BFile::get_max_min(vector<real> &max, vector<real> &min)
{
	seek_position(0);

	vector<real> R(rvars_,0);
	vector<integer> D(dvars_,0);
	int ID;

	real bignum = 10000000;

	// initialize max and min vectors 
	for (int i=0;i<R.size();i++) {
		min[i] = bignum;
		max[i] = -bignum;
	}

	// process rest of examples
	while (index_ < records_) {
		get_next(ID,R,D);

		for (int i=0;i<R.size();i++) {
			if (R[i] != missing_r_) {
				if (R[i] < min[i]) {
					min[i] = R[i];
				}
				else if (R[i] > max[i]) {
					max[i] = R[i];
				}
			}
		};
	}

};



//--------------------------------------------------------------------
// get_mean_std
//
// Makes a pass through the data and calculates the mean and standard
// deviation for real variables.
//--------------------------------------------------------------------
void BFile::get_mean_std(vector<real> &mean, vector<real> &std)
{
	seek_position(0);

	vector<real> R(rvars_,0);
	vector<integer> D(dvars_,0);
	int ID;

	vector<double> sumv(rvars_,0);
	vector<double> sumsqv(rvars_,0);
	vector<int> num(rvars_,0);

	// process rest of examples
	while (index_ < records_) {
		get_next(ID,R,D);

		for (int i=0;i<R.size();i++) {
			if (R[i] != missing_r_) {
				double r = ((double) R[i]);
				sumv[i] += r;
				sumsqv[i] += r*r;
				num[i]++;
			}
		};
	}

	cout << "  sum: ";
	printvector(sumv);
	cout << endl;
	cout << "  sum^2: ";
	printvector(sumsqv);
	cout << endl;

	for (int i=0;i<rvars_;i++) {
		if (num[i] > 1) {
			double mean_value = sumv[i] / num[i];
			mean[i] = ((real) mean_value);

			double std_value = sqrt((sumsqv[i] - sumv[i] * sumv[i] / num[i]) / (num[i] - 1));
			std[i] = ((real) std_value);
		}
		else {
			mean[i] = 0;
			std[i] = 0;
		}
		// error checkin
		// check mean /std are not NaN
	}
};



//--------------------------------------------------------------------
// multi_shuffle
//--------------------------------------------------------------------
void BFile::multi_shuffle(string destfile, int iterations, int tmpfiles)
{
	assert(tmpfiles > 0);
	for (int i=0;i<iterations;i++) {
		shuffle(destfile,10000,tmpfiles);
		set_file_ptr(destfile);
	};
};


//--------------------------------------------------------------------
// shuffle
//
// file
// blocksize
// ntmpfiles
//--------------------------------------------------------------------
void BFile::shuffle(string file, int blocksize, int ntmpfiles)
{

	//-------------------------
	// set up tmp file names
	//
	vector<string> tmpfiles;
	for (int i=0;i<ntmpfiles;i++) {
		//ostringstream filename;
		strstream filename;
		filename << "tmp." << i << ends;
		//cout << "filename: " << filename.str() << endl;
		//cout << filename.str() << endl;
		tmpfiles.push_back(filename.str());
	}


	//-------------------------------
	// open file for writing
	//

	vector<ofstream*> file_ptrs;
	for (int i=0;i<tmpfiles.size();i++) {
		ofstream *outfile = new ofstream(tmpfiles[i].c_str(),ios::out|ios::binary);

		if ( ! *outfile ) {
			cerr << "Error: cannot open file " << file << endl;
			exit(0);
		}

		file_ptrs.push_back(outfile);
	};


	//--------------------------------
	// read in data file and randomly shuffle examples to
	// temporay files
	//
	cout << "  Splitting data set randomly into temporary files" << endl;
	seek_position(0);

	vector<real> R(rvars_,0);
	vector<integer> D(dvars_,0);
	int ID;
	while (index_ < records_) {
		get_next(ID,R,D);
		
		// for linux gcc version
		//double rnum = floor(drand48() * ((double) tmpfiles.size()));

		// for MINGW version
		double drand = (((double) rand()) / ((double) RAND_MAX+1));
		double rnum = floor(drand * ((double) tmpfiles.size()));

//		cout << drand << "   " << rnum << endl;

		int index = ((int) rnum); 
		
		file_ptrs[index]->write((const char *) &ID,sizeof(int));	
		file_ptrs[index]->write((const char *) &R[0],sizeof(real)*rvars_);	
		file_ptrs[index]->write((const char *) &D[0],sizeof(integer)*dvars_);	
	};

	// close temporary file pointers and delete ofstream objects
	for (int i=0;i<file_ptrs.size();i++) {
		file_ptrs[i]->close();
		delete file_ptrs[i];
	}


	//-------------------------------
	// open tmpfiles for reading 
	//

	vector<ifstream*> ifile_ptrs;
	for (int i=0;i<tmpfiles.size();i++) {
		ifstream *infile = new ifstream(tmpfiles[i].c_str(), ios::in | ios::binary);

		if ( ! *infile) {
			cerr << "Error: cannot open file " << file << endl;
			exit(0);
		}

		ifile_ptrs.push_back(infile);
	};


	//-----------------------------------
	// open final destination file
	//
	
	ofstream outfile (file.c_str(),ios::out|ios::binary);
	if ( ! outfile ) {
		cerr << "Error: cannot open file " << file << endl;
		exit(0);
	};
	
	// write header information
	outfile.write((const char *) &records_,sizeof(int));
	outfile.write((const char *) &rvars_,sizeof(int));
	outfile.write((const char *) &dvars_,sizeof(int));


	//--------------------------------------
	// concatenate tmp files in random order
	// 

	vector<int> order;
	for (int i=0;i<tmpfiles.size();i++) {
		order.push_back(i);
	}
	random_shuffle(order.begin(),order.end());
	cout << "  Combining temporary files in order: ";
	printvector(order);
	cout << endl;
	

	for (int i=0;i<order.size();i++) {
		int count = blocksize;
		ifstream *infile = ifile_ptrs[order[i]];
		while(count == blocksize) {
			//char buffer[blocksize];
			char *buffer = new char[blocksize];
			infile->read(buffer,blocksize);
			count = infile->gcount();
			outfile.write((const char *) buffer,count);
			delete [] buffer;
		}
	}
	
	
	//-----------------------------
	// close file pointers
	//
	
	outfile.close();

	// close temporary file pointers and delete ifstream objects
	for (int i=0;i<file_ptrs.size();i++) {
		ifile_ptrs[i]->close();
		delete ifile_ptrs[i];
	}

	//-----------------------------
	// clean temporary files
	//
	for (int i=0;i<tmpfiles.size();i++) {
		if (remove(tmpfiles[i].c_str()) == -1) {
			cerr << "Error deleting file " << tmpfiles[i] << endl;
		}
	}
};




//--------------------------------------------------------------------
// get_next 
//
// WARNING does not check to make sure read command succeeded.
//--------------------------------------------------------------------
void BFile::get_next(int &ID, vector<real> &R, vector<integer> &D)
{
	
	infile_->read((char *) &ID, sizeof(int));
	infile_->read((char *) &R[0], sizeof(real) * rvars_);
	infile_->read((char *) &D[0], sizeof(integer) * dvars_);

	index_++;
};

//--------------------------------------------------------------------
// set_file_ptr
//--------------------------------------------------------------------
void BFile::set_file_ptr(string file)
{
	// if infile_ already points to a file, close it and
	// free the memory for the ifstream object
	if (infile_ != 0) {
		infile_->close();
		delete infile_;
	}

	// create a new ifstream object
  infile_ = new ifstream(file.c_str(), ios::in | ios::binary);
	index_ = 0;
	if ( ! infile_) {
		cerr << "error: unabile to open input file: " << file << endl;
		exit(-1);
	}
};



//--------------------------------------------------------------------
// read_header
//--------------------------------------------------------------------
void BFile::read_header(void)
{
	// check file ptr != 0
	int data[3];
	infile_->read((char *) data,3*sizeof(int));

	records_ = data[0];
	rvars_ = data[1];
	dvars_ = data[2];

};


//--------------------------------------------------------------------
// seek_position
//--------------------------------------------------------------------
void BFile::seek_position(int pos)
{
	// check if position valie
	assert(pos >= 0);
	assert(pos < records_);
				
	//int filepos = 3 * sizeof(int) + pos * sizeof(real) + pos * sizeof(integer);
	long filepos = 3 * sizeof(int) + pos * sizeof(real) + pos * sizeof(integer);
	infile_->seekg(filepos);
	index_ = pos;
				
};

//--------------------------------------------------------------------
// print_fields
//--------------------------------------------------------------------
void BFile::print_header(void)
{
	cout << "Records: " << records_ << endl;
	cout << "Real Variables: " << rvars_ << endl;
	cout << "Discrete Variables: " << dvars_ << endl;
};


//--------------------------------------------------------------------
// print_records
//--------------------------------------------------------------------
void BFile::print_records(int num)
{
	seek_position(0);
	
	vector<real> R(rvars_,0);
	vector<integer> D(dvars_,0);
	int ID;
	if (num == -1) {
		num = records_;
	}

	int count = 0;
	while ((index_ < records_) && (count < num)) {
		get_next(ID,R,D);
		
		cout << ID << " ";
		printvector(R);
		cout << " | " ;
		printvector(D);
		cout << endl;
		count++;
	}
	
};

//--------------------------------------------------------------------
// RStats
//--------------------------------------------------------------------
RStats::RStats(int size)
{
	vector<real> zero(size,0);
	max_ = zero;
	min_ = zero;
	mean_ = zero;
	std_ = zero;
};


//--------------------------------------------------------------------
// load
//--------------------------------------------------------------------
void RStats::load(string file)
{
	// open file
	fstream infile(file.c_str(), ios::in);
	if (!infile) {
		cerr << "error: unable to open input file: " << file << endl;
		exit(0);
	}
	
	string buffer;

	vector<string> delimiters;
	delimiters.push_back(":");
	delimiters.push_back(";");
	delimiters.push_back(",");

	while (getline(infile,buffer)) {
		vector<string> tokens = tokenize(buffer,delimiters);

		// if we have a valid line
		vector<real> *V;
		V = &std_;
		if (tokens.size() > 0) {
			if (tokens[0] == "max") 
				V = &max_;
			else if (tokens[0] == "min")
				V = &min_;
			else if (tokens[0] == "mean")
				V = &mean_;
			else if (tokens[0] == "std")
				V = &std_;
			else {
				cerr << "error: unknown statistic " << tokens[0] << " in file " << file << endl;
				exit(0);
			}
		}
				
		// get numbers
		for (int i=1; i<tokens.size();i++) {
			//V->push_back(atof(tokens[i].c_str()));
			(*V)[i-1] = atof(tokens[i].c_str());
		}
 	}

};


//--------------------------------------------------------------------
// print
//--------------------------------------------------------------------
void RStats::print(int which) 
{
	if (which == 0) {
		if (max_.size() != 0) {
			cout << "  max: ";
			printvector(max_);
			cout << endl;
		}
		if (min_.size() != 0) {
			cout << "  min: ";
			printvector(min_);
			cout << endl;
		}
	}
	else {
		if (mean_.size() != 0) {
			cout << "  mean: ";
			printvector(mean_);
			cout << endl;
		}
		if (std_.size() != 0) {
			cout << "  std: ";
			printvector(std_);
			cout << endl;
		}
	}
};

File: Source/Orca/DPrep/DPrep_Sharp/BFile.cs
﻿using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Thesis.DPrep
{
    class BFile : IDisposable
    {
        BinaryReader _infile;
        int _example;

        string _dataFile;

        int _index;

        int _records;
        int _realFieldsCount;
        int _discreteFieldsCount;

        float _missingR;
        int _missingD;

        public BFile(string filename, float missingR, int missingD)
        {
            _dataFile = filename;
            _missingR = missingR;
            _missingD = missingD;

            SetFileReader(filename);
            ReadHeader();
        }

        private void SetFileReader(string filename)
        {
            // if infile_ already points to a file, close it
            if (_infile != null)
                _infile.Close();

            _infile = new BinaryReader(File.OpenRead(filename));
            _index = 0;
        }

        private void ReadHeader()
        {
            _records = _infile.ReadInt32();
            _realFieldsCount = _infile.ReadInt32();
            _discreteFieldsCount = _infile.ReadInt32();
        }

        private void SeekPosition(int pos)
        {
            if (pos < 0 || pos > _records)
                throw new ArgumentOutOfRangeException("pos");

            long filepos = 3 * sizeof(int) + pos * sizeof(float) + pos * sizeof(int);
            _infile.BaseStream.Seek(pos, SeekOrigin.Begin);
            _index = pos;
        }

        private void GetNext(out int id, out float[] real, out int[] discrete)
        {
            // WARNING does not check to make sure read command succeeded.
            id = _infile.ReadInt32();
            real = new float[_realFieldsCount];
            discrete = new int[_discreteFieldsCount];
            byte[] temp = _infile.ReadBytes(sizeof(float) * _realFieldsCount);
            Buffer.BlockCopy(temp, 0, real, 0, temp.Length);
            temp = _infile.ReadBytes(sizeof(int) * _discreteFieldsCount);
            Buffer.BlockCopy(temp, 0, discrete, 0, temp.Length);
            
            _index++;
        }

        /// <summary>
        /// Makes a pass through the data and calculates the max and min values
        /// for real variables.
        /// </summary>
        //public void GetMaxMin(ref IList<float> max, ref IList<float> min)
        public void GetMaxMin(float[] max, float[] min)
        {
            SeekPosition(0);

            //if (max == null)
            //    max = new List<float>(_realFieldsCount);
            //if (min == null)
            //    min = new List<float>(_realFieldsCount);

            //var R = new List<float>(RealFieldsCount);
            //var D = new List<int>(DiscreteFieldsCount);

            // initialize max and min vectors 
            for (int i = 0; i < _realFieldsCount; i++)
            {
                min[i] = float.MaxValue;
                max[i] = float.MinValue;
            }

            // process rest of examples
            while (_index < _records)
            {
                int id;
                float[] R;
                int[] D;
                GetNext(out id, out R, out D);

                for (int i = 0; i < _realFieldsCount; i++)
                {
                    if (R[i] != _missingR)
                    {
                        if (R[i] < min[i])
                        {
                            min[i] = R[i];
                        }
                        else if (R[i] > max[i])
                        {
                            max[i] = R[i];
                        }
                    }
                };
            }
        }

        /// <summary>
        /// Makes a pass through the data and calculates the mean and standard
        /// deviation for real variables.
        /// </summary>
        public void GetMeanStd(float[] mean, float[] std)
        {
            SeekPosition(0);
            var sumv = new double[_realFieldsCount];
            var sumsqv = new double[_realFieldsCount];
            var num = new int[_realFieldsCount];

            while (_index < _records)
            {
                int id;
                float[] R;
                int[] D;
                GetNext(out id, out R, out D);

                for (int i = 0; i < _realFieldsCount; i++)
                {
                    if (R[i] != _missingR)
                    {
                        double r = ((double)R[i]);
                        sumv[i] += r;
                        sumsqv[i] += r * r;
                        num[i]++;
                    }
                }

                for (int i = 0; i < _realFieldsCount; i++)
                {
                    if (num[i] > 1)
                    {
                        double meanValue = sumv[i] / num[i];
                        mean[i] = ((float)meanValue);

                        double stdValue = Math.Sqrt((sumsqv[i] - sumv[i] * sumv[i] / num[i]) / (num[i] - 1));
                        std[i] = ((float)stdValue);
                    }
                    else
                    {
                        mean[i] = 0;
                        std[i] = 0;
                    }
                    // error checkin
                    // check mean /std are not NaN
                }
            }
        }

        /// <summary>
        /// Scales real values to the range [0,1].
        /// </summary>
        public void ScaleZeroToOne(string filename, float[] max, float[] min)
        {
            //-------------------------------
	        // open file for writing
	        //
            using (var outstream = File.Create(filename))
            using (BinaryWriter outfile = new BinaryWriter(outstream))
            {
	            // write header information
                outfile.Write(_records);
                outfile.Write(_realFieldsCount);
                outfile.Write(_discreteFieldsCount);

	            float[] range = new float[_realFieldsCount];

	            for (int i = 0; i < _realFieldsCount; i++) {
		            range[i] = max[i] - min[i];
	            };

	            //--------------------------------
	            // read in file and scale it
	            //
	            SeekPosition(0);

	            float[] Rscale = new float[_realFieldsCount];
                while (_index < _records)
                {
                    int id;
                    float[] R;
                    int[] D;
                    GetNext(out id, out R, out D);

                    for (int i = 0; i < _realFieldsCount; i++)
                    {
                        if (R[i] == _missingR)
                        {
                            Rscale[i] = _missingR;
                        }
                        else if (range[i] != 0)
                        {
                            Rscale[i] = (R[i] - min[i]) / range[i];
                        }
                        else
                        {
                            Rscale[i] = 0;
                        }
                    }

                    outfile.Write(id);
                    outfile.Write(Rscale);
                    outfile.Write(D);
                }

                outfile.Close();
	        }
        }

        /// <summary>
        /// Scales real values to standard deviations from the mean.
        /// </summary>
        public void ScaleStd(string filename, float[] mean, float[] std)
        {
            //-------------------------------
            // open file for writing
            //
            using (var outstream = File.Create(filename))
            using (BinaryWriter outfile = new BinaryWriter(outstream))
            {
                // write header information
                outfile.Write(_records);
                outfile.Write(_realFieldsCount);
                outfile.Write(_discreteFieldsCount);

                //--------------------------------
                // read in file and scale it
                //
                SeekPosition(0);

                float[] Rscale = new float[_realFieldsCount];
                while (_index < _records)
                {
                    int id;
                    float[] R;
                    int[] D;
                    GetNext(out id, out R, out D);

                    for (int i = 0; i < _realFieldsCount; i++)
                    {
                        if (R[i] == _missingR)
                        {
                            Rscale[i] = _missingR;
                        }
                        else if (std[i] != 0)
                        {
                            Rscale[i] = (R[i] - mean[i]) / std[i];
                        }
                        else
                        {
                            Rscale[i] = 0;
                        }
                    }

                    outfile.Write(id);
                    outfile.Write(Rscale);
                    outfile.Write(D);
                }

                outfile.Close();
            }
        }

        public void MultiShuffle(string destFile, int iterations, int tmpFiles, int seed)
        {
            if (tmpFiles <= 0)
                throw new ArgumentOutOfRangeException("tmpfiles");

            Random rand = new Random(seed);
            for (int i = 0; i < iterations; i++)
            {
                Shuffle(destFile, 10000, tmpFiles, rand);
                SetFileReader(destFile);
            }
        }

        private void Shuffle(string file, int blockSize, int nTmpFiles, Random rand)
        {
            //-------------------------
            // set up tmp file names
            //
            string[] tmpFileNames = new string[nTmpFiles];
            for (int i = 0; i < nTmpFiles; i++)
                tmpFileNames[i] = file + ".tmp." + i.ToString();


            //-------------------------------
            // open files for writing
            //
            BinaryWriter[] tmpFilesOut = new BinaryWriter[nTmpFiles];
            try
            {
                for (int i = 0; i < tmpFileNames.Length; i++)
                    tmpFilesOut[i] = new BinaryWriter(File.Create(tmpFileNames[i]));

                //--------------------------------
                // read in data file and randomly shuffle examples to
                // temporary files
                //
                SeekPosition(0);

                while (_index < _records)
                {
                    int id;
                    float[] R;
                    int[] D;
                    GetNext(out id, out R, out D);
                    int index = rand.Next(tmpFilesOut.Length);

                    tmpFilesOut[index].Write(id);
                    tmpFilesOut[index].Write(R);
                    tmpFilesOut[index].Write(D);
                }
            }
            finally
            {
                // close temporary files
                for (int i = 0; i < tmpFilesOut.Length; i++)
                    if (tmpFilesOut[i] != null)
                        tmpFilesOut[i].Dispose();
            }

            //-------------------------------
            // open tmpfiles for reading 
            //
            
            BinaryReader[] tmpFilesIn = new BinaryReader[nTmpFiles];
            try
            {
                for (int i = 0; i < tmpFilesIn.Length; i++)
                    tmpFilesIn[i] = new BinaryReader(File.OpenRead(tmpFileNames[i]));

                //-----------------------------------
                // open final destination file
                //

                using (var outstream = File.Create(file))
                using (var outfile = new BinaryWriter(outstream))
                {
                    outfile.Write(_records);
                    outfile.Write(_realFieldsCount);
                    outfile.Write(_discreteFieldsCount);

                    //--------------------------------------
                    // concatenate tmp files in random order
                    //
                    int[] order = new int[nTmpFiles];
                    for (int i = 0; i < nTmpFiles; i++)
                        order[i] = i;

                    // The modern version of the Fisher–Yates shuffle (the Knuth shuffle)
                    for (int i = order.Length; i > 0; i--)
                    {
                        int j = rand.Next(i + 1);
                        int temp = order[i];
                        order[i] = order[j];
                        order[j] = temp;
                    }

                    for (int i = 0; i < order.Length; i++)
                    {
                        int count = blockSize;
                        BinaryReader infile = tmpFilesIn[order[i]];
                        while (count == blockSize)
                        {
                            byte[] temp = infile.ReadBytes(blockSize);
                            count = temp.Length;
                            outfile.Write(temp);
                        }
                    }

                    outfile.Close();
                }
            }
            finally
            {
                // close temporary files
                for (int i = 0; i < tmpFilesIn.Length; i++)
                    if (tmpFilesIn[i] != null)
                        tmpFilesIn[i].Dispose();
            }
        }

        public int RecordsCount
        {
            get
            {
                return _records;
            }
        }

        public int RealFieldsCount
        {
            get
            {
                return _realFieldsCount;
            }
        }

        public int DiscreteFieldsCount
        {
            get
            {
                return _discreteFieldsCount;
            }
        }
        
        #region IDisposable
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
 
        private bool m_Disposed = false;
 
        protected virtual void Dispose(bool disposing)
        {
            if (!m_Disposed)
            {
                if (disposing)
                {
                // Managed resources are released here.
                    _infile.Close();
                }
 
                // Unmanaged resources are released here.
                m_Disposed = true;
            }
        }
 
        ~BFile()    
        {        
            Dispose(false);
        }
        #endregion
    }
}
